### PR TITLE
Allow concurrency to be set with percentage values

### LINKED ISSUE #761: Ability to set concurrency to max (equal to CPU cores) `--concurrency=max`
### Describe the feature you'd like to request

I am not an expert in this area, but I think it's counter-productive to set the concurrency to a higher number than there are CPU cores available for CPU-hungry processes, like tests and builds, as these processes would be competing for resources.

Thus, a default setting of 10 might be counter-productive in certain environments, like CI, where the number of cores can be lower.

For example, [GitHub Actions](https://docs.github.com/en/actions/using-github-hosted-runners/about-github-hosted-runners#supported-runners-and-hardware-resources) only have 2 cores available for Windows and Linux, yet 3 for macOS.

### Describe the solution you'd like

I think there should be a way to specify "max" or some other variation of that, that signals Turbo to sniff out the number of cores and set the concurrency to that value.

### Describe alternatives you've considered

```sh
turbo run build --concurrency=$(node -r os -e 'process.stdout.write(os.cpus().length.toString())')
```

### PR DESCRIPTION
**Pull request by VanTanev**
Closes #761

This is modeled after the way that `jest` [allows to set workers by providing a percent with `--maxWorkers=50%`](https://jestjs.io/docs/configuration#maxworkers-number--string).

I've updated Turbo's CLI parsing to accept `--concurrency=50%`. It also supports percentages over 100%, so  `--concurrency=500%` is valid. It uses `runtime.NumCPU()` to get the number of available logical processors, and calculates from there.

Setting the percentage to zero or less is an error. Setting percentage very low, to eg `1%` results in always at least `1` as the value for concurrency.

### CODE REVIEW

**Comment by VanTanev** [comment]
I would also want to work on adding concurrency to the `turbo.json` config file but wanted to submit one PR at a time.

**Comment by VanTanev** [comment]
@jaredpalmer Is there anything I can do to help move this PR forward? Any changes I should do?

**Review by gsoltis — CHANGES_REQUESTED** [review]
Overall this looks solid. Just one minor nit.

File: cli/internal/util/parse_concurrency.go [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+
+var (
+	// alias so we can mock in tests
+	runtimeNumCPU = runtime.NumCPU
+)
+
+func ParseConcurrency(concurrencyRaw string) (int, error) {
+	if strings.HasSuffix(concurrencyRaw, "%") {
+		if percent, err := strconv.ParseFloat(concurrencyRaw[:len(concurrencyRaw)-1], 64); err != nil {
+			return -1, fmt.Errorf("invalid value for --concurrency CLI flag. This should be a number --concurrency=4 or percentage of CPU cores --concurrency=50%% : %w", err)
```
[the remark this answers is not in the document]
Comment (VanTanev) [reply cli/internal/util/parse_concurrency.go]: Fixed, thanks!

**Comment by gsoltis** [comment]
@VanTanev Maybe start a discussion for the config file? `turbo.json` is intended to be checked in and shared, and it's not immediately clear to me that everyone would want to default to the same concurrency. Just as an example, I would imagine your CI might want 100% while devs might prefer some headroom left over. My intuition is that this is a setting that is per-machine, like login information. But happy to get input from the community on what would be helpful here.

**Review by gsoltis — APPROVED** [review]

**Comment by moltar** [comment]
@VanTanev Thanks for this PR!

IMO having an environment config would be more useful that having an entry in the config file.

To me, `turbo.json` config is something permanent, that is project-dependant, and does not change from environment to environment.

But what changes from environment to environment are ... the environment variables 😁 

I might want to run at 100% on CI, and at 50% on my local machine, because it's busy doing other things.

And that's what the environment variables are exactly for! I could even configure my local env permanently via bashrc to _always_ run turbo on 50% settings (again, because I want to use my dev box while it's running builds, regardless of the project). But I _always_ want to take up 100% of the CI.

And if turbo is just another command in a custom build script, then I lose control of the concurrency setting via CLI option and I also cannot change the config from env to env.

**Comment by VanTanev** [comment]
@moltar Can I add both, with precedence going argument > env > turbo.json?

**Comment by moltar** [comment]
> Can I add both

@VanTanev I am not the project maintainer, merely the one who raised issue #761 😁 

But I personally see no value in having it inside the config. But to each their own!

**Pull request merged**